Add CustomStateSet and :state() selector

keithamus opened the pull request:
#### 59cb7cb2d9244ba77870eabe9f6a8da9249a1312
<pre>
Add CustomStateSet and :state() selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=215911">https://bugs.webkit.org/show_bug.cgi?id=215911</a>

Reviewed by Tim Nguyen.

This adds the CustomStateSet idl definitions, and the `states` getter in
the ElementInternals IDL. The gatter laziliy initialises a
CustomStateSet object in ElementRareData which behaves like a
setlike&lt;DOMString&gt;.

This also adds the `:state()` pseudo selector, which can take an ident
and matches when the ident matches one of the given states in the
CustomStateSet for the element.

Any states set by CustomStateSet need to invalidate styles for the
element, so the `:state()` pseudo can match.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/state-in-has-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/state-in-has.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/parsing/parse-state-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/ElementInternals-states-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/ElementInternals-states.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/custom-state-set-strong-ref.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/idlharness.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/state-css-selector-shadow-dom.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/state-css-selector-shadow-dom.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/state-css-selector.tentative-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isOnlyPseudoClassFunction):
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/dom/CustomStateSet.cpp: Added.
(WebCore::CustomStateSet::addToSetLike):
(WebCore::CustomStateSet::removeFromSetLike):
(WebCore::CustomStateSet::clearFromSetLike):
(WebCore::CustomStateSet::has):
* Source/WebCore/dom/CustomStateSet.h: Added.
* Source/WebCore/dom/CustomStateSet.idl: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasCustomState const):
(WebCore::Element::ensureCustomStateSet):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::ElementInternals::states):
* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/dom/ElementInternals.idl:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::customStateSet):
(WebCore::ElementRareData::setCustomStateSet):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/NodeRareData.h:

Canonical link: <a href="https://commits.webkit.org/271600@main">https://commits.webkit.org/271600@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e495b9ddeba51a3f18adf54a843558ef6b9d1911

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28971 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30327 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31585 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26396 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29558 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9759 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4959 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29243 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6307 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5485 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/25885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32923 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26487 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31861 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29243 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5586 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29639 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7238 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/25679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/35586 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6916 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6079 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7673 "Passed tests") | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -684,6 +685,10 @@ String CSSSelector::selectorText(StringView separator, StringView rightSide) con
+            case CSSSelector::PseudoClassType::State:
+                builder.append(":state(");
+                appendPseudoClassFunctionTail(builder, cs);
+                break;
```
```suggestion
            case CSSSelector::PseudoClassType::State:
                builder.append(":state(");
                appendPseudoClassFunctionTail(builder, cs);
                builder.append(')');
                break;
```

nt1m replied:
Would be good to have parsing WPT (See LayoutTests/imported/w3c/web-platform-tests/css/selectors/parsing). That WPT could also cover invalid cases like `:state`

nt1m replied:
This doesn't seem resolved, `builder.append(')');` is missing.

keithamus replied:
I don't think this is necessary, because `appendPseudoClassFunctionTail` calls:

```cxx
        builder.append(selector->argument(), ')');
```

Unless I'm misunderstanding the `append` function, in which case please let me know and I'll happily add the additional line.

nt1m replied:
Ah right, I was mistaken with another part of the code, sorry.

nt1m commented on Source/WebCore/dom/CustomStateSet.h:
```diff
@@ -0,0 +1,61 @@
+    {
+    }
+
+    Vector<const AtomString*> m_states;
```
I would probably use a ListHashSet here

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5513,4 +5514,31 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
+    if (!document().settings().customStateSetEnabled())
+        return false;
+
+    if(hasRareData()) {
```
```suggestion
    if (hasRareData()) {
```

nt1m commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,68 @@
+    m_states.append(&state_atom);
+    Element* element = m_element.get();
+    if (element && !removed)
+        element->invalidateStyleForCustomStateChange();
```
I don't think this would actually invalidate. The PseudoClassInvalidationChange needs to be in the same scope as when the matching changes. You'll want something among those lines:
```suggestion
    m_states.append(&state_atom);
    Element* element = m_element.get();
    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
    if (element && !removed)
        styleInvalidation.emplace(m_element, CSSSelector::PseudoClassType::State, Style::PseudoClassChangeInvalidation::AnyValue);
```

nt1m replied:
Also it would be good to have invalidation WPT for this. Simple invalidation and also invalidation with `:has()`.

See:
LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/defined-in-has.html
LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/defined.html

nt1m commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,68 @@
+bool CustomStateSet::addToSetLike(String state)
+{
+    const AtomString state_atom = AtomString(state);
+    bool removed = m_states.removeFirst(&state_atom);
```
The spec seems to refer to this: https://infra.spec.whatwg.org/#set-append (though I might be wrong)

So I think there's actually no re-ordering to do if the state is already there.

nt1m commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,68 @@
+    bool removed = m_states.removeFirst(&state_atom);
+    Element* element = m_element.get();
+    if (element && removed)
+        element->invalidateStyleForCustomStateChange();
```
ditto

nt1m commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,68 @@
+    m_states.clear();
+    Element* element = m_element.get();
+    if (element)
+        element->invalidateStyleForCustomStateChange();
```
ditto

nt1m reviewed (commented):
You'll need to add a special case at: https://searchfox.org/wubkat/rev/023c54054092dc68c5df3b230ed3137cbd753b16/Source/WebCore/css/parser/CSSSelectorParser.cpp#743-763

Otherwise, `:state` with no argument won't be rejected.

rniwa commented on Source/WebCore/css/SelectorChecker.cpp:
```diff
@@ -1087,6 +1087,9 @@ bool SelectorChecker::checkOne(CheckingContext& checkingContext, const LocalCont
             const Node* contextualReferenceNode = !checkingContext.scope ? element.document().documentElement() : checkingContext.scope;
             return &element == contextualReferenceNode;
         }
+        case CSSSelector::PseudoClassType::State: {
```
Curly braes is not needed here.

rniwa commented on Source/WebCore/cssjit/SelectorCompiler.cpp:
```diff
@@ -446,6 +446,7 @@ struct SelectorFragment {
     Vector<SelectorList> matchesFilters;
     Vector<Vector<SelectorFragment>> anyFilters;
     const CSSSelector* pseudoElementSelector = nullptr;
+    Vector<const AtomString*> customStateSet;
```
Please use `Vector<AtomString>` instead.

nt1m replied:
This doesn't seem resolved.

rniwa commented on Source/WebCore/dom/CustomStateSet.idl:
```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
```
2023?

rniwa commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5513,4 +5514,31 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
+CustomStateSet* Element::ensureCustomStateSet()
+{
+    ElementRareData& rareData = const_cast<Element*>(this)->ensureElementRareData();
+    if (!rareData.customStateSet()) {
```
No curly braces around a single line statement.

nt1m commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,71 @@
+    if (element)
+        styleInvalidation.emplace(*element, CSSSelector::PseudoClassType::State, Style::PseudoClassChangeInvalidation::AnyValue);
+
+    return m_states.add(AtomString(state)).isNewEntry;;
```
```suggestion
    return m_states.add(AtomString(state)).isNewEntry;
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/selectors/parsing/parse-state.html:
```diff
@@ -0,0 +1,16 @@
+  test_valid_selector(":state(--0)");
+  test_valid_selector(":host(:state(--foo))");
+  test_valid_selector('my-input[type="foo"]:state(checked)');
+  test_valid_selector('my-input[type="foo"]:state(--0)::before');
```
I would add:
```
test_invalid_selector(":state");
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/custom-elements/state.tentative/state-pseudo-class.html:
Can we do a WPT resync of custom-elements/ and custom-state-pseudo-class separately ? I think it'll make it easier for you to export, and it's easier to review generally since it's clear which ones are new tests.

nt1m commented on Source/WebCore/cssjit/SelectorCompiler.cpp:
```diff
@@ -1337,6 +1338,10 @@ static inline FunctionType addPseudoClassType(const CSSSelector& selector, Selec
+    case CSSSelector::PseudoClassType::State:
+        fragment.customStateSet.append(selector.argument());
+        return FunctionType::SimpleSelectorChecker;
+
```
Right now, we're populating this data and not using it. This is supposed to be used for CSSJIT, see: https://github.com/WebKit/WebKit/commit/57ad5510bfa05b8679d85c9becf7819e62854a1e

You should either implement CSSJIT support fully or remove the `fragment.customStateSet` field and set the pseudo class to `FunctionType::CannotCompile`

keithamus replied:
Ah I thought this was sufficient, but now I see it's a little more complex I'll skip it for this PR and perhaps come back to add it in another PR.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class-expected.txt:
```diff
@@ -1,9 +1,9 @@
+PASS :state(foo) serialization
+PASS :state(foo) in simple cases
+PASS :state(foo) and other pseudo classes
+FAIL :state(foo) and ::part() assert_equals: ::part() followed by :state(--foo) expected "0.5" but got "0"
```
This can be fixed in a followup, but it is surprising that this doesn't work. `::part()` does support pseudo classes in WebKit.

nt1m commented on Source/WebCore/css/parser/CSSSelectorParserContext.h:
```diff
@@ -43,6 +43,7 @@ struct CSSSelectorParserContext {
     bool highlightAPIEnabled { false };
     bool popoverAttributeEnabled { false };
     bool viewTransitionsEnabled { false };
+    bool customStateSetEnabled { false };
```
This needs to be alpha sorted if you want it to build :)

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class.html:
```diff
@@ -106,27 +112,27 @@
-      '::part() followed by :--foo');
-  innerStates.delete('--innerFoo');
-  innerStates.add('--innerfoo');
+      '::part() followed by :state(--foo)');
```
Can you change to `'::part() followed by :state(innerFoo)'` or `'::part() followed by :state(foo)'`?
One of the two for consistency

cdumez commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,72 @@
+{
+    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
+    Element* element = m_element.get();
+    if (element)
```
`if (RefPtr element = m_element.get())`

cdumez commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,72 @@
+{
+    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
+    Element* element = m_element.get();
+    if (element)
```
ditto.

cdumez commented on Source/WebCore/dom/CustomStateSet.idl:
```diff
@@ -0,0 +1,31 @@
+    Exposed=Window,
+    EnabledBySetting=CustomStateSetEnabled,
+] interface CustomStateSet {
+    setlike<DOMString>;
```
Have you tried using setlike<[AtomString]DOMString> and then have removeFromSetLike() and other take in a `const AtomString&` instead of a `const String&`? Since their implementation seems to want an AtomString.

I am not 100% sure it'd work but it is worth trying.

cdumez commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,72 @@
+{
+    std::optional<Style::PseudoClassChangeInvalidation> styleInvalidation;
+    Element* element = m_element.get();
+    if (element)
```
ditto.

cdumez commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,72 @@
+    m_states.clear();
+}
+
+bool CustomStateSet::has(AtomString state)
```
`const AtomString&`

cdumez commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,72 @@
+    return m_states.add(AtomString(state)).isNewEntry;
+}
+
+bool CustomStateSet::removeFromSetLike(String state)
```
`const String&` or ideally `const AtomString&` since it seems to be what you need inside the implementation.

cdumez commented on Source/WebCore/dom/CustomStateSet.cpp:
```diff
@@ -0,0 +1,72 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(CustomStateSet);
+
+bool CustomStateSet::addToSetLike(String state)
```
`const String&` or ideally `const AtomString&` since it seems to be what you need inside the implementation.

cdumez commented on Source/WebCore/dom/CustomStateSet.h:
```diff
@@ -0,0 +1,61 @@
+    void clearFromSetLike();
+    void initializeSetLike(DOMSetAdapter&) { };
+
+    bool has(AtomString);
```
can the member function be marked as `const`?

cdumez commented on Source/WebCore/dom/CustomStateSet.h:
```diff
@@ -0,0 +1,61 @@
+    bool has(AtomString);
+
+private:
+    CustomStateSet(Element& element)
```
please mark as `explicit`.

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5531,4 +5532,25 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
+        return false;
+
+    if (hasRareData()) {
+        CustomStateSet* customStates = elementRareData()->customStateSet();
```
`CustomStateSet*` -> `RefPtr`

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5531,4 +5532,25 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
+
+    if (hasRareData()) {
+        CustomStateSet* customStates = elementRareData()->customStateSet();
+        return customStates ? customStates->has(*state) : false;
```
`return customStates && customStates->has(*state);`

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5531,4 +5532,25 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
+
+CustomStateSet* Element::ensureCustomStateSet()
+{
+    ElementRareData& rareData = const_cast<Element*>(this)->ensureElementRareData();
```
auto&

cdumez commented on Source/WebCore/dom/ElementInternals.cpp:
```diff
@@ -191,4 +192,9 @@ void ElementInternals::setElementsArrayAttribute(const QualifiedName& name, std:
 
+CustomStateSet* ElementInternals::states()
+{
+    return m_element.get()->ensureCustomStateSet();
```
`m_element.get()->` -> `m_element->`

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5531,4 +5532,25 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
     return target;
 }
 
+bool Element::hasCustomState(const AtomString* state) const
```
Why does this take a pointer? Seems it should take a `const AtomString&`.

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5531,4 +5532,25 @@ AtomString Element::makeTargetBlankIfHasDanglingMarkup(const AtomString& target)
+    return false;
+}
+
+CustomStateSet* Element::ensureCustomStateSet()
```
Why does this return a raw pointer? It doesn't seem it can return null. If so, this should return a `CustomStateSet&`.

nt1m commented on Source/WebCore/dom/CustomStateSet.idl:
```diff
@@ -0,0 +1,31 @@
+    Exposed=Window,
+    EnabledBySetting=CustomStateSetEnabled,
+] interface CustomStateSet {
+    setlike<[AtomString]DOMString>;
```
```suggestion
    setlike<[AtomString] DOMString>;
```
This is the style the rest of our code uses.

keithamus committed:
Add CustomStateSet and :state() selector
https://bugs.webkit.org/show_bug.cgi?id=215911

Reviewed by Tim Nguyen.

This adds the CustomStateSet idl definitions, and the `states` getter in
the ElementInternals IDL. The gatter laziliy initialises a
CustomStateSet object in ElementRareData which behaves like a
setlike<DOMString>.

This also adds the `:state()` pseudo selector, which can take an ident
and matches when the ident matches one of the given states in the
CustomStateSet for the element.

Any states set by CustomStateSet need to invalidate styles for the
element, so the `:state()` pseudo can match.

* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/state-in-has-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/invalidation/state-in-has.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/parsing/parse-state-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/ElementInternals-states-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/ElementInternals-states.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-elements/state/tentative/state-pseudo-class.html:
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/custom-state-set-strong-ref.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/idlharness.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/state-css-selector-shadow-dom.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/state-css-selector-shadow-dom.tentative.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/custom-state-pseudo-class/state-css-selector.tentative-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorPseudoClassAndCompatibilityElementMap.in:
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::isOnlyPseudoClassFunction):
(WebCore::CSSSelectorParser::consumePseudo):
* Source/WebCore/css/parser/CSSSelectorParserContext.cpp:
(WebCore::CSSSelectorParserContext::CSSSelectorParserContext):
(WebCore::add):
* Source/WebCore/css/parser/CSSSelectorParserContext.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):
* Source/WebCore/dom/CustomStateSet.cpp: Added.
(WebCore::CustomStateSet::addToSetLike):
(WebCore::CustomStateSet::removeFromSetLike):
(WebCore::CustomStateSet::clearFromSetLike):
(WebCore::CustomStateSet::has):
* Source/WebCore/dom/CustomStateSet.h: Added.
* Source/WebCore/dom/CustomStateSet.idl: Added.
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::hasCustomState const):
(WebCore::Element::ensureCustomStateSet):
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementInternals.cpp:
(WebCore::ElementInternals::states):
* Source/WebCore/dom/ElementInternals.h:
* Source/WebCore/dom/ElementInternals.idl:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::customStateSet):
(WebCore::ElementRareData::setCustomStateSet):
(WebCore::ElementRareData::useTypes const):
* Source/WebCore/dom/NodeRareData.h:

Canonical link: https://commits.webkit.org/271600@main
